File: leet_code/easy/21_merge_two_sorted_lists.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        if l1 is None and l2 is None:
            return None
        if l1 is None:
            return l2
        if l2 is None:
            return l1

        head = None
        node = None
        node_1 = l1
        node_2 = l2
        while True:
            if node is None:
                if node_1.val <= node_2.val:
                    head = node_1
                    node_1 = node_1.next
                else:
                    head = node_2
                    node_2 = node_2.next

                node = head
            elif node_1 is None:
                node.next = node_2
                break
            elif node_2 is None:
                node.next = node_1
                break
            else:
                if node_1.val <= node_2.val:
                    node.next = node_1
                    node_1 = node_1.next
                else:
                    node.next = node_2
                    node_2 = node_2.next

                node = node.next

        return head

File: leetcode/easy/21_merge_two_sorted_lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

// -----------------------------------------
// My Solution
//
// Time  Complexity: O(m + n)
// Space Complexity: O(1)
// -----------------------------------------
// m := number of nodes of l1, n := number of nodes of l2
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        if (l1 == NULL) return l2;
        if (l2 == NULL) return l1;

        // initialize new head of merged list
        ListNode* new_head;
        if (l1->val < l2->val) {
            new_head = l1;
            l1 = l1->next;
        } else {
            new_head = l2;
            l2 = l2->next;
        }

        // setup remaining nodes in while loop
        ListNode* curr_node = new_head;
        while (l1 != NULL || l2 != NULL) {
            if (l1 == NULL) {
                curr_node->next = l2;
                break;
            }
            if (l2 == NULL) {
                curr_node->next = l1;
                break;
            }

            if (l1->val < l2->val) {
                curr_node->next = l1;
                l1 = l1->next;
            } else {
                curr_node->next = l2;
                l2 = l2->next;
            }
            curr_node = curr_node->next;
        }

        return new_head;
    }
};
